daemon: use netlink for managed neighbor support probe

rgo3 opened the pull request:
With this commit we refactor the probe logic that checks for the existence of a certain helper as this could break on older distro kernels that have backported the commit we probed for.

Using actual netlink calls to probe for managed neighbor support will only succeed if the underlying kernel actually supports it.

Fixes #20694.

rgo3 committed:
daemon: use netlink for managed neighbor support probe

With this commit we refactor the probe logic that checks for
the existence of a certain helper as this could break on older
distro kernel that have backported the commit we probed for.

Using actual netlink calls to probe for managed neighbor support
will only succeed if the underlying kernel actually supports it.

Fixes #20694.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

lmb commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -361,20 +363,62 @@ func probeKubeProxyReplacementOptions() error {
 }
 
-func probeManagedNeighborSupport() {
+func probeManagedNeighborSupport() error {
```
Nit: take `Config` as a parameter and write a unit test for the function?

rgo3 replied:
To just test the happy path (no errors) with a privileged test?

lmb replied:
Yeah pretty much!

ti-mo committed:
testutils: add SkipOnOldKernel test helper for feature probes

This helper allows tests for feature probes to be skipped on kernel versions
that are known to lack the feature.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -361,20 +363,62 @@ func probeKubeProxyReplacementOptions() error {
+	}
+
+	if err := netns.Set(oldns); err != nil {
+		return fmt.Errorf("failed to switch back to original netns '%s': %w", oldns.UniqueId(), err)
```
`defer runtime.UnlockOSThread()` will cause the OS thread to get unlocked even if setting the thread back to the initial netns fails (or if any of the errors above fire, really). This is really bad, as the thread will be released back into the schedulable pool, but remain in the empty netns. Same as with bpf objects; `newns` is simply a handle, and the active thread maintains a reference to the netns.

`runtime.UnlockOSThread()` may only be called after a successful call to netns.Set(). If that doesn't succeed, the only safe way to proceed is by letting the OS thread die, which means terminating the current goroutine without calling UnlockOSThread().

In order to do this safely, we'll need to start `probeManagedNeighborSupport` in a goroutine, so we can make sure the goroutine ends when the function exits, giving us control over the lifecycle of the underlying thread.

Additionally, the logic will need to be changed to this:

```go
defer func() {
  // defer closes over named return variable err

  if nerr := netns.Set(oldns); nerr != nil {
    // The current goroutine is locked to an OS thread and we've failed
    // to undo state modifications to the thread. 
    err = fmt.Errorf("error setting thread back to its original netns: %w (original error: %s)", nerr, err)
    return
  }

  runtime.UnlockOSThread()
}
```

ti-mo commented on daemon/cmd/daemon.go:
```diff
@@ -415,7 +415,18 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
+		case errors.Is(err, probes.ErrNotSupported):
+			log.Info("Managed Neigbors are not supported by the kernel")
+		case err != nil:
+			log.WithError(err).Warn("Unable to probe managed neighbor support")
```
`probes.HaveProgramHelper` makes probe failures fatal, I think we should keep the same principle here. I find warnings confusing, as they're actually errors that fly under the radar. Does this require user intervention?

rgo3 replied:
Felt like a harsh intervention to fatal the agent for a failed probe that just determines how something is handled internally in the kernel (or does this enable any special features?), but I see that just warning about it doesn't really improve anything either. If we want to fatal I can move the error checking to the wrapper function `HaveManagedNeighbors` and fatal there as with the other probe APIs and simplify the error handling for the caller.

rgo3 replied:
I agree that we should keep the same principal, to keep things consistent. If a probe fails inconclusively something weird is going on.

ti-mo commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/vishvananda/netns"
+)
+
+// HaveManagedNeighbors creates a netlink.Neigh entry with the flag NTF_EXT_MANAGED.
```
This needs to highlight the function's semantics:
```suggestion
// HaveManagedNeighbors returns nil if the host supports managed neighbor entries (NTF_EXT_MANAGED).
// Any errors encountered during the feature probe are returned verbatim.
```

ti-mo commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+
+// HaveManagedNeighbors creates a netlink.Neigh entry with the flag NTF_EXT_MANAGED.
+func HaveManagedNeighbors() error {
+	ch := make(chan bool)
```
```suggestion
	ch := make(chan struct{})
```

ti-mo commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+
+	// In order to call HaveManagedNeighbors safely, it has to be started
+	// in a goroutine, so we can make sure the goroutine ends when the function exits.
+	// This gives us control over the lifecycle of the underlying thread.
```
```suggestion
	// This makes sure the underlying OS thread exits if we fail to restore it to the original netns.
```

ti-mo commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+// HaveManagedNeighbors creates a netlink.Neigh entry with the flag NTF_EXT_MANAGED.
+func HaveManagedNeighbors() error {
```
`err` can be stored in a package global gated by sync.Once so it's fast to call repeatedly.

ti-mo reviewed (changes requested):
Looks good! Just a few nits.

tklauser commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+func haveManagedNeighbors() (outer error) {
+	runtime.LockOSThread()
```
I think we'll want to `defer runtime.UnlockOSThread()` here instead of further down in the `defer func() { ... }` otherwise we won't unlock in case of an error in `netns.{Get,New}()` below.

ti-mo replied:
Hmm, that's by design. If we defer the unlock here, we can't stop the OS thread from becoming schedulable again if netns.Set fails. Using Goexit instead of return wouldn't help either, since it runs all deferred functions before terminating the goroutine.

netns.Get/New is so unlikely to fail (Cilium targets container systems after all) that I don't consider letting the OS thread die here to be a big deal. The probe results are supposed to be cached, so no risk of calling it in a loop and causing impact.

tklauser replied:
Thanks for the explanation. I just noticed https://github.com/cilium/cilium/pull/25134#discussion_r1189953611 now which explains the rationale behind this. I think this should at least be briefly mentioned in a comment so others reading the code later on don't stumble upon this.

ti-mo replied:
Another option is calling `runtime.LockOSThread(); defer runtime.UnlockOSThread()` here and adding an extra `LockOSThread()` after checking the error from `netns.New()`.

In order for an OS thread to be unlocked, as many calls to `Unlock()` need to be made as there were calls to `Lock()`.

rgo3 replied:
I left it the way it was, but added a comment to explain the late `defer runtime.UnlockOSThread()`.

tklauser commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+	// This gives us control over the lifecycle of the underlying thread.
+	go func() {
+		err = haveManagedNeighbors()
+		ch <- true
```
IMO it's slightly more idiomatic to just close the channel:

```suggestion
		close(ch)
```

lmb commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,84 @@
+			// The current goroutine is locked to an OS thread and we've failed
+			// to undo state modifications to the thread.
+			outer = fmt.Errorf("error setting thread back to its original netns: %w (original error: %s)", nerr, outer)
+			return
```
Good catch with not doing UnlockOSThread when this fails, but shouldn't we call https://pkg.go.dev/runtime#Goexit ?

ti-mo replied:
~In this particular case, return and goexit have the same effect, yes. Calling goexit instead may make our intention clearer and prevent future changes from silently introducing a bug, agreed.~ Not sure why I said this. If we Goexit(), `outer` will be lost and `HaveManagedNeighbors()` will hang on the read from `<-ch`.

rgo3 replied:
How do I ensure that the goroutine closes the channel if I call Goexit in `haveManagedNeighbors` and don't end up with main in a deadlock? Would I have to do something like this?

```golang
go func() {
    defer close(ch)
    managedNeighborResult = haveManagedNeighbors()
}()
```
The documentation says "Goexit runs all deferred calls before terminating the goroutine".

Also if we call Goexit, is outer even returned from `haveManagedNeighbors()`?

ti-mo replied:
Sorry, that was sleep deprivation talking. Of course this needs to `return`.. If we want to use goexit here, we'd have to send the return value to the creating goroutine using a channel or something. Let's stick with `return`.

ti-mo commented on pkg/datapath/linux/probes/managed_neighbors.go:
```diff
@@ -0,0 +1,104 @@
+)
+
+// HaveManagedNeighbors returns nil if the host supports managed neighbor entries (NTF_EXT_MANAGED).
+// Any errors encountered during the feature probe are returned verbatim.
```
This is no longer true. Let's add the same note as on the other Have...() functions:
```suggestion
// On unexpected probe results this function will terminate with log.Fatal().
```

ti-mo reviewed (approved):
LGTM! One last nit on the docstring.

tklauser reviewed (approved):
Nice, thanks for adding the explanatory comment in `haveManagedNeighbors`.
